File: JS/29 Course schedule - naive BFS.js
const p = [
  [1, 0],
  [2, 1],
  [2, 5],
  [0, 3],
  [4, 3],
  [3, 5],
  [4, 5],
];

const canFinish = function (n, prerequisites) {
  // generate empty structure of adjancency list
  const adjList = new Array(n).fill(0).map(() => []);

  // fill adjancency list from preequisites list
  for (let i = 0; i < prerequisites.length; i++) {
    // get the pair from the prerequisites list
    const pair = prerequisites[i];

    // get the second value from the pair
    // and push first value from pair
    // inside adjlist
    // adjList[0].push(1);
    adjList[pair[1]].push(pair[0]);
  }

  for (let v = 0; v < n; v++) {
    // queue to store vertex
    const queue = [];
    // seen to track vertex we visited
    const seen = {};

    for (let i = 0; i < adjList[v].length; i++) {
      // push value from the adj list to queue
      queue.push(adjList[v][i]);
    }

    // BFS
    while (queue.length) {
      // get the current value from queue
      const current = queue.shift();
      // store this current value in seen object
      seen[current] = true;

      //Test case, check cycle
      if (current === v) return false;

      // get the current value
      const adjacent = adjList[current];

      //
      for (let i = 0; i < adjacent.length; i++) {
        // get the next value
        const next = adjacent[i];
        //check, as long we havent seen value before then,
        if (!seen[next]) {
          // store this in queue
          queue.push(next);
        }
      }
    }
  }

  return true;
};

console.log(canFinish(6, p));

File: Refactor/Python Thonny/29 Course schedule - naive BFS.py
p = [
     [1, 0], 
     [2, 1], 
     [2, 5], 
     [0, 3], 
     [4, 3], 
     [3, 5], 
     [4, 5]
    ]

def canFinish(n, prerequisites):
    #  generate empty structure of adjancency list
    adjList = list(map(lambda i : [] , [None] * n))
    #adjList = new Array(n).fill(0).map(() => []) JS code
  

  # fill adjancency list from preequisites list
    for i in range(len(prerequisites)):
        # get the pair from the prerequisites list
        pair = prerequisites[i]

        # get the second value from the pair
        # and push first value from pair
        # inside adjlist
        # adjList[0].push(1);
        adjList[pair[1]].append(pair[0])
  

    for v in range(n):
        # queue to store vertex
        queue = []
        # seen dictionary to track vertex we visited
        seen = {}

        for i in range(len(adjList[v])):
            # push value from the adj list to queue
            queue.append(adjList[v][i])
    
        #  BFS
        while(len(queue)):
            # get the current value from queue
            # const current = queue.shift();
            current = queue.pop(0)
            # store this current value in seen dictionary
            seen[current] = True

            # Test case, check cycle
            if(current == v):
                return False

            # get the current value
            adjacent = adjList[current]
        
            for i in range(len(adjacent)):   
                # get the next value   
                next = adjacent[i]
                # check, as long we havent seen value before then,
                if(not seen.get(next)):
                    # store this in queue
                    queue.append(next)
    
    return True


print(canFinish(6, p))
